File: trunk/ace/Stream.cpp
// Stream.cpp
// $Id$

#if !defined (ACE_STREAM_C)
#define ACE_STREAM_C

#define ACE_BUILD_DLL
//#include "ace/Module.h"
#include "ace/Stream.h"
#include "ace/Stream_Modules.h"

#if !defined (__ACE_INLINE__)
#include "ace/Stream.i"
#endif /* __ACE_INLINE__ */

ACE_ALLOC_HOOK_DEFINE(ACE_Stream)

// Give some idea of what the heck is going on in a stream!

template <ACE_SYNCH_1> void
ACE_Stream<ACE_SYNCH_2>::dump (void) const
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::dump");
  ACE_DEBUG ((LM_DEBUG, "-------- module links --------\n"));

  for (ACE_Module<ACE_SYNCH_2> *mp = this->stream_head_; 
       ; 
       mp = mp->next ())
    {
      ACE_DEBUG ((LM_DEBUG, "module name = %s\n", mp->name ()));
      if (mp == this->stream_tail_)
	break;
    }

  ACE_DEBUG ((LM_DEBUG, "-------- writer links --------\n"));

  ACE_Task<ACE_SYNCH_2> *tp;

  for (tp = this->stream_head_->writer (); 
       ; 
       tp = tp->next ())
    {
      ACE_DEBUG ((LM_DEBUG, "writer queue name = %s\n", tp->name ()));
      tp->dump ();
      ACE_DEBUG ((LM_DEBUG, "-------\n"));
      if (tp == this->stream_tail_->writer ()
	  || (this->linked_us_ 
	      && tp == this->linked_us_->stream_head_->reader ()))
	break;
    }

  ACE_DEBUG ((LM_DEBUG, "-------- reader links --------\n"));
  for (tp = this->stream_tail_->reader (); ; tp = tp->next ())
    {
      ACE_DEBUG ((LM_DEBUG, "reader queue name = %s\n", tp->name ()));
      tp->dump ();
      ACE_DEBUG ((LM_DEBUG, "-------\n"));
      if (tp == this->stream_head_->reader ()
	  || (this->linked_us_ 
	      && tp == this->linked_us_->stream_head_->writer ()))
	break;
    }
}

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::push (ACE_Module<ACE_SYNCH_2> *new_top)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::push");
  if (this->push_module  (new_top, 
			  this->stream_head_->next (), 
			  this->stream_head_) == -1)
    return -1;
  else
    return 0;
}

template <ACE_SYNCH_1> int
ACE_Stream_Iterator<ACE_SYNCH_2>::advance (void)
{
  ACE_TRACE ("ACE_Stream_Iterator<ACE_SYNCH_2>::advance");
  this->next_ = this->next_->next ();
  return 0;
}

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::put (ACE_Message_Block *mb, ACE_Time_Value *tv)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::put");
  return this->stream_head_->writer ()->put (mb, tv);
}

template <ACE_SYNCH_1> int 
ACE_Stream<ACE_SYNCH_2>::get (ACE_Message_Block *&mb, ACE_Time_Value *tv)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::get");
  return this->stream_head_->reader ()->getq (mb, tv);
}

// Return the "top" ACE_Module in a ACE_Stream, skipping over the
// stream_head.

template <ACE_SYNCH_1> int 
ACE_Stream<ACE_SYNCH_2>::top (ACE_Module<ACE_SYNCH_2> *&m)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::top");
  if (this->stream_head_->next () == this->stream_tail_)
    return -1;
  else
    {
      m = this->stream_head_->next ();
      return 0;
    }
}

// Remove the "top" ACE_Module in a ACE_Stream, skipping over the
// stream_head.

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::pop (int flags)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::pop");
  if (this->stream_head_->next () == this->stream_tail_)
    return -1;
  else
    {
      // Skip over the ACE_Stream head.
      ACE_Module<ACE_SYNCH_2> *top_mod = this->stream_head_->next ();
      ACE_Module<ACE_SYNCH_2> *new_top = top_mod->next ();

      this->stream_head_->next (new_top);

      // Close the top ACE_Module.

      top_mod->close (flags);

      // Don't delete the Module unless the flags request this.
      if (flags != ACE_Module<ACE_SYNCH_2>::M_DELETE_NONE)
	delete top_mod;

      this->stream_head_->writer ()->next (new_top->writer ());
      new_top->reader ()->next (this->stream_head_->reader ());
      return 0;
    }
}

// Remove a named ACE_Module from an arbitrary place in the
// ACE_Stream.

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::remove (const char *name, 
				 int flags)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::remove");
  ACE_Module<ACE_SYNCH_2> *prev = 0;

  for (ACE_Module<ACE_SYNCH_2> *mod = this->stream_head_; 
       mod != 0; 
       mod = mod->next ())
    if (ACE_OS::strcmp (mod->name (), name) == 0)
      {
	if (prev == 0) // Deleting ACE_Stream Head
	  this->stream_head_->link (mod->next ());
	else
	  prev->link (mod->next ());

	// Don't delete the Module unless the flags request this.
	if (flags != ACE_Module<ACE_SYNCH_2>::M_DELETE_NONE)
	  {
	    // Close down the module and release the memory.
	    mod->close (flags);
	    delete mod;
	  }

	return 0;
      }
    else
      prev = mod;

  return -1;
}

template <ACE_SYNCH_1> ACE_Module<ACE_SYNCH_2> *
ACE_Stream<ACE_SYNCH_2>::find (const char *name)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::find");
  for (ACE_Module<ACE_SYNCH_2> *mod = this->stream_head_; 
       mod != 0; 
       mod = mod->next ())
    if (ACE_OS::strcmp (mod->name (), name) == 0)
	return mod;

  return 0;
}

// Actually push a module onto the stack...

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::push_module (ACE_Module<ACE_SYNCH_2> *new_top, 
				      ACE_Module<ACE_SYNCH_2> *current_top, 
				      ACE_Module<ACE_SYNCH_2> *head)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::push_module");
  ACE_Task<ACE_SYNCH_2> *nt_reader = new_top->reader ();
  ACE_Task<ACE_SYNCH_2> *nt_writer = new_top->writer ();
  ACE_Task<ACE_SYNCH_2> *ct_reader = 0;
  ACE_Task<ACE_SYNCH_2> *ct_writer = 0;

  if (current_top)
    {
      ct_reader = current_top->reader ();
      ct_writer = current_top->writer ();
      ct_reader->next (nt_reader);
    }

  nt_writer->next (ct_writer);

  if (head)
    {
      if (head != new_top)
	head->link (new_top);
    }
  else
    nt_reader->next (0);

  new_top->next (current_top);

  if (nt_reader->open (new_top->arg ()) == -1)
    return -1;

  if (nt_writer->open (new_top->arg ()) == -1)
    return -1;
  return 0;
}

#if 0
int
ACE_Stream<ACE_SYNCH_2>::open (void *a, 
			       ACE_Multiplexor &muxer, 
			       ACE_Module<ACE_SYNCH_2> *head)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::open");
  this->stream_head_ = head == 0
    ? new ACE_Module<ACE_SYNCH_2> ("ACE_Stream_Head", 
			       new ACE_Stream_Head<ACE_SYNCH_2>, 
			       new ACE_Stream_Head<ACE_SYNCH_2>, a) : head;
  this->stream_tail_ = 0;
  return muxer.link_from_below (this->stream_head_);
}
#endif 

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::open (void *a, 
			       ACE_Module<ACE_SYNCH_2> *head, 
			       ACE_Module<ACE_SYNCH_2> *tail)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::open");
  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->lock_, -1);

  ACE_Task<ACE_SYNCH_2> *h1 = 0, *h2 = 0;
  ACE_Task<ACE_SYNCH_2> *t1 = 0, *t2 = 0;

  if (head == 0)
    {
      h1 = new ACE_Stream_Head<ACE_SYNCH_2>;
      h2 = new ACE_Stream_Head<ACE_SYNCH_2>;
      head = new ACE_Module<ACE_SYNCH_2> ("ACE_Stream_Head", h1, h2, a);
    }

  if (tail == 0)
    {
      t1 = new ACE_Stream_Tail<ACE_SYNCH_2>;
      t2 = new ACE_Stream_Tail<ACE_SYNCH_2>;
      tail = new ACE_Module<ACE_SYNCH_2> ("ACE_Stream_Tail", 
					  t1, t2, a);
    }

  // Make sure *all* the allocation succeeded!
  if (head == 0 && (h1 == 0 || h2 == 0) 
      || tail == 0 && (t1 == 0 || t2 == 0))
    {
      delete h1;
      delete h2;
      delete t1;
      delete t2;
      delete head;
      delete tail;
      errno = ENOMEM;
      return -1;
    }

  this->stream_head_ = head;
  this->stream_tail_ = tail;

  if (this->push_module (this->stream_tail_) == -1)
    return -1;
  else if (this->push_module (this->stream_head_, 
			      this->stream_tail_, 
			      this->stream_head_) == -1)
    return -1;
  else
    return 0;
}

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::close (int flags)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::close");
  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->lock_, -1);

  if (this->stream_head_ != 0 
      && this->stream_tail_ != 0)
    {
      // Don't bother checking return value here.
      this->unlink_i ();

      int result = 0;

      // Remove and cleanup all the intermediate modules.

      while (this->stream_head_->next () != this->stream_tail_)
	if (this->pop (flags) == -1)
	  result = -1;
  
      // Clean up the head and tail of the stream.
      if (this->stream_head_->close (flags) == -1)
	result = -1;
      if (this->stream_tail_->close (flags) == -1)
	result = -1;

      // Cleanup the memory.
      delete this->stream_head_;
      delete this->stream_tail_;

      this->stream_head_ = 0;
      this->stream_tail_ = 0;

      // Tell all threads waiting on the close that we are done.
      this->final_close_.broadcast ();
      return result;
    }
  return 0;
}

template <ACE_SYNCH_1> int 
ACE_Stream<ACE_SYNCH_2>::control (ACE_IO_Cntl_Msg::ACE_IO_Cntl_Cmds cmd, 
				  void *a)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::control");
  ACE_IO_Cntl_Msg ioc (cmd);

  // Create a data block that contains the user-supplied data.
  ACE_Message_Block *db = 
    new ACE_Message_Block (sizeof (int), 
			   ACE_Message_Block::MB_IOCTL, 
			   0, 
			   (char *) a);

  // Create a control block that contains the control field and a
  // pointer to the data block.
  ACE_Message_Block *cb = 
    new ACE_Message_Block (sizeof ioc, 
			   ACE_Message_Block::MB_IOCTL, 
			   db, 
			   (char *) &ioc);

  // Make sure all of the allocation succeeded before continuing.
  if (db == 0 || cb == 0)
    {
      delete cb; 
      delete db; 
      errno = ENOMEM;
      return -1;
    }

  int result = 0;

  if (this->stream_head_->writer ()->put (cb) == -1)
    result = -1;
  else if (this->stream_head_->reader ()->getq (cb) == -1)
    result = -1;
  else
    result = ((ACE_IO_Cntl_Msg *) cb->rd_ptr ())->rval ();

  delete cb; // This also deletes db...
  return result;
}

// Link two streams together at their bottom-most Modules (i.e., the
// one just above the Stream tail).  Note that all of this is premised
// on the fact that the Stream head and Stream tail are non-NULL...
// This must be called with locks held.

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::link_i (ACE_Stream<ACE_SYNCH_2> &us)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::link_i");
  this->linked_us_ = &us;
  // Make sure the other side is also linked to us!
  us.linked_us_ = this;

  ACE_Module<ACE_SYNCH_2> *my_tail = this->stream_head_;

  if (my_tail == 0)
    return -1;

  // Locate the module just above our Stream tail. 
  while (my_tail->next () != this->stream_tail_)
    my_tail = my_tail->next ();

  ACE_Module<ACE_SYNCH_2> *other_tail = us.stream_head_;

  if (other_tail == 0)
    return -1;

  // Locate the module just above the other Stream's tail.
  while (other_tail->next () != us.stream_tail_)
    other_tail = other_tail->next ();

  // Reattach the pointers so that the two streams are linked!
  my_tail->writer ()->next (other_tail->reader ());
  other_tail->writer ()->next (my_tail->reader ());
  return 0;
}

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::link (ACE_Stream<ACE_SYNCH_2> &us)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::link");
  
  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->lock_, -1);

  return this->link_i (us);
}  

// Must be called with locks held...

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::unlink_i (void)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::unlink_i");

  // Only try to unlink if we are in fact still linked!

  if (this->linked_us_ != 0)
    {
      ACE_Module<ACE_SYNCH_2> *my_tail = this->stream_head_;

      // Only relink if we still exist!
      if (my_tail)
	{
	  // Find the module that's just before our stream tail.
	  while (my_tail->next () != this->stream_tail_)
	    my_tail = my_tail->next ();

	  // Restore the writer's next() link to our tail.
	  my_tail->writer ()->next (this->stream_tail_->writer ());
	}

      ACE_Module<ACE_SYNCH_2> *other_tail = 
	this->linked_us_->stream_head_;

      // Only fiddle with the other side if it in fact still remains.
      if (other_tail != 0)
	{
	  while (other_tail->next () != this->linked_us_->stream_tail_)
	    other_tail = other_tail->next ();

	  other_tail->writer ()->next (this->linked_us_->stream_tail_->writer ());

	}

      // Make sure the other side is also aware that it's been unlinked!
      this->linked_us_->linked_us_ = 0;

      this->linked_us_ = 0;
      return 0;
    }
  else
    return -1;
}

template <ACE_SYNCH_1> int
ACE_Stream<ACE_SYNCH_2>::unlink (void)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::unlink");
  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->lock_, -1);
  return this->unlink_i ();
}

template <ACE_SYNCH_1>
ACE_Stream<ACE_SYNCH_2>::ACE_Stream (void * a, 
				     ACE_Module<ACE_SYNCH_2> *head, 
				     ACE_Module<ACE_SYNCH_2> *tail)
  : linked_us_ (0),
    final_close_ (this->lock_)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::ACE_Stream");
  if (this->open (a, head, tail) == -1)
    ACE_ERROR ((LM_ERROR, "ACE_Stream<ACE_SYNCH_2>::open (%s, %s)\n", 
	       head->name (), tail->name ()));
}

#if 0
ACE_Stream<ACE_SYNCH_2>::ACE_Stream (void *a, 
				     ACE_Multiplexor &muxer, 
				     ACE_Module<ACE_SYNCH_2> *head)
  : linked_us_ (0),
    final_close_ (this->lock_)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::ACE_Stream");
  if (this->open (a, muxer, head) == -1)
    ACE_ERROR ((LM_ERROR, "ACE_Stream<ACE_SYNCH_2>::open (%s, %s)\n", 
		head->name ()));
}
#endif 

template <ACE_SYNCH_1>
ACE_Stream<ACE_SYNCH_2>::~ACE_Stream (void)
{
  ACE_TRACE ("ACE_Stream<ACE_SYNCH_2>::~ACE_Stream");

  if (this->stream_head_ != 0)
    this->close ();
}

template <ACE_SYNCH_1>
ACE_Stream_Iterator<ACE_SYNCH_2>::ACE_Stream_Iterator (const ACE_Stream<ACE_SYNCH_2> &sr)
  : next_ (sr.stream_head_)
{
  ACE_TRACE ("ACE_Stream_Iterator<ACE_SYNCH_2>::ACE_Stream_Iterator");
}

#endif /* ACE_STREAM_C */

File: trunk/java/src/Stream.java
/*************************************************
 *
 * = PACKAGE
 *    ACE.ASX
 *
 * = FILENAME
 *    Stream.java
 *
 *@author Prashant Jain
 *
 *************************************************/
package ACE.ASX;

import ACE.OS.*;

/**
 * <hr>
 * <h2>SYNOPSIS</h2>
 *<blockquote>
 *     This class is the primary abstraction for the ASX framework.
 *     It is moduled after System V Stream.
 *</blockquote>
 *
 * <h2>DESCRIPTION</h2>
 *<blockquote>
 *   A Stream consists of a stack of <Modules>, each of which
 *   contains two <Tasks>.
 *</blockquote>
 */

public class Stream
{

  public Stream ()
  {
    this (null, null, null);
  }

  // Create a Stream consisting of <head> and <tail> as the Stream
  // head and Stream tail, respectively.  If these are 0 then the
  // <ACE_Stream_Head> and <ACE_Stream_Tail> are used, respectively.
  // <arg> is the value past in to the open() methods of the tasks.

  public Stream (Object a,
		 Module head,
		 Module tail)
  {
    this.linkedUs_ = null;
    // this.final_close_ = this.lock_;

    if (this.open (a, head, tail) == -1)
      ACE.ERROR ("open" + head.name () + " " +  tail.name ());
  }

  public int push (Module newTop)
  {
    if (this.pushModule  (newTop, 
			  this.streamHead_.next (), 
			  this.streamHead_) == -1)
      return -1;
    else
      return 0;
  }

  public int put (MessageBlock mb, TimeValue tv)
  {
    return this.streamHead_.writer ().put (mb, tv);
  }

  public MessageBlock get (TimeValue tv) throws InterruptedException
  {
    return this.streamHead_.reader ().getq (tv);
  }

// Return the "top" ACE_Module in a ACE_Stream, skipping over the
// stream_head.

  public Module top ()
  {
    if (this.streamHead_.next () == this.streamTail_)
      return null;
    else
      return this.streamHead_.next ();
  }

// Remove the "top" ACE_Module in a ACE_Stream, skipping over the
// stream_head.

  public int pop (long flags)
  {
    if (this.streamHead_.next () == this.streamTail_)
      return -1;
    else
      {
	// Skip over the ACE_Stream head.
	Module top = this.streamHead_.next ();
	Module newTop = top.next ();

	this.streamHead_.next (newTop);

	// Close the top ACE_Module.

	top.close (flags);

	this.streamHead_.writer ().next (newTop.writer ());
	newTop.reader ().next (this.streamHead_.reader ());

	return 0;
      }
  }

// Remove a named ACE_Module from an arbitrary place in the
// ACE_Stream.

  public int remove (String name, long flags)
  {
    Module prev = null;

    for (Module mod = this.streamHead_; 
	 mod != null; mod = mod.next ())
      if (name.compareTo (mod.name ()) == 0)
	{
	  if (prev == null) // Deleting ACE_Stream Head
	    this.streamHead_.link (mod.next ());
	  else
	    prev.link (mod.next ());
	  
	  mod.close (flags);
	  return 0;
	}
      else
	prev = mod;

    return -1;
  }

  public Module find (String name)
  {
    for (Module mod = this.streamHead_; 
	 mod != null; 
	 mod = mod.next ())
      if (name.compareTo (mod.name ()) == 0)
	return mod;

    return null;
  }

// Actually push a module onto the stack...

  private int pushModule (Module newTop, 
			  Module currentTop, 
			  Module head)
  {
    Task ntReader = newTop.reader ();
    Task ntWriter = newTop.writer ();
    Task ctReader = null;
    Task ctWriter = null;

    if (currentTop != null)
      {
      ctReader = currentTop.reader ();
      ctWriter = currentTop.writer ();
      ctReader.next (ntReader);
      }

    ntWriter.next (ctWriter);

    if (head != null)
      {
	if (head != newTop)
	  head.link (newTop);
      }
    else
      ntReader.next (null);

    newTop.next (currentTop);

    if (ntReader.open (newTop.arg ()) == -1)
      return -1;

    if (ntWriter.open (newTop.arg ()) == -1)
      return -1;
    return 0;
  }

  public synchronized int open (Object a, 
				Module head, 
				Module tail)
  {
    Task h1 = null, h2 = null;
    Task t1 = null, t2 = null;

    if (head == null)
      {
	h1 = new StreamHead ();
	h2 = new StreamHead ();
	head = new Module ("ACEStreamHead", h1, h2, a);
      }

    if (tail == null)
      {
	t1 = new StreamTail ();
	t2 = new StreamTail ();
	tail = new Module ("ACEStreamTail", 
			   t1, t2, a);
      }

    // Make sure *all* the allocation succeeded!
    if (h1 == null || h2 == null || head == null
	|| t1 == null || t2 == null || tail == null)
      {
      // Close up!
	head.close (0);
	tail.close (0);
	return -1;
      }

    this.streamHead_ = head;
    this.streamTail_ = tail;

    if (this.pushModule (this.streamTail_,
			 null, null) == -1)
      return -1;
    else if (this.pushModule (this.streamHead_, 
			      this.streamTail_, 
			      this.streamHead_) == -1)
      return -1;
    else
      return 0;
  }

  public synchronized int close (long flags)
  {
    if (this.streamHead_ != null
	&& this.streamTail_ != null)
      {
	// Don't bother checking return value here.
	this.unlinkInternal ();

	int result = 0;

	// Remove and cleanup all the intermediate modules.

	while (this.streamHead_.next () != this.streamTail_)
	  {
	    if (this.pop (flags) == -1)
	      result = -1;
	  }
  
	// Clean up the head and tail of the stream.
	if (this.streamHead_.close (flags) == -1)
	  result = -1;
	if (this.streamTail_.close (flags) == -1)
	  result = -1;

	this.streamHead_ = null;
	this.streamTail_ = null;

	// Tell all threads waiting on the close that we are done.
	//      this.final_close_.broadcast ();
	return result;
      }
    return 0;
  }

  public int control (int cmd, Object a) throws InterruptedException
  {
    IOCntlMsg ioc = new IOCntlMsg (cmd);

    // Create a data block that contains the user-supplied data.
    MessageBlock db = 
      new MessageBlock (MessageType.MB_IOCTL, 
			null, 
			a);

    // Create a control block that contains the control field and a
    // pointer to the data block.
    MessageBlock cb = 
      new MessageBlock (MessageType.MB_IOCTL, 
			db, 
			(Object) ioc);

    int result = 0;

    if (this.streamHead_.writer ().put (cb, new TimeValue ()) == -1)
      result = -1;
    else if ((cb = this.streamHead_.reader ().getq (new TimeValue ())) == null)
      result = -1;
    else
      result = ((IOCntlMsg ) cb.obj ()).rval ();

    return result;
  }

// Link two streams together at their bottom-most Modules (i.e., the
// one just above the Stream tail).  Note that all of this is premised
// on the fact that the Stream head and Stream tail are non-NULL...
// This must be called with locks held.

  private int linkInternal (Stream us)
  {
    this.linkedUs_ = us;
    // Make sure the other side is also linked to us!
    us.linkedUs_ = this;

    Module myTail = this.streamHead_;

    if (myTail == null)
      return -1;

    // Locate the module just above our Stream tail. 
    while (myTail.next () != this.streamTail_)
      myTail = myTail.next ();

    Module otherTail = us.streamHead_;

    if (otherTail == null)
      return -1;

  // Locate the module just above the other Stream's tail.
    while (otherTail.next () != us.streamTail_)
      otherTail = otherTail.next ();

    // Reattach the pointers so that the two streams are linked!
    myTail.writer ().next (otherTail.reader ());
    otherTail.writer ().next (myTail.reader ());
    return 0;
  }

  public synchronized int link (Stream us)
  {
    return this.linkInternal (us);
  }  

// Must be called with locks held...

  private int unlinkInternal ()
  {
  // Only try to unlink if we are in fact still linked!
    
    if (this.linkedUs_ != null)
    {
      Module myTail = this.streamHead_;

      // Only relink if we still exist!
      if (myTail != null)
	{
	  // Find the module that's just before our stream tail.
	  while (myTail.next () != this.streamTail_)
	    myTail = myTail.next ();

	  // Restore the writer's next() link to our tail.
	  myTail.writer ().next (this.streamTail_.writer ());
	}

      Module otherTail = this.linkedUs_.streamHead_;

      // Only fiddle with the other side if it in fact still remains.
      if (otherTail != null)
	{
	  while (otherTail.next () != this.linkedUs_.streamTail_)
	    otherTail = otherTail.next ();

	  otherTail.writer ().next (this.linkedUs_.streamTail_.writer ());
	  
	}

      // Make sure the other side is also aware that it's been unlinked!
      this.linkedUs_.linkedUs_ = null;

      this.linkedUs_ = null;
      return 0;
    }
    else
      return -1;
  }

  public synchronized int unlink ()
  {
    return this.unlinkInternal ();
  }

  public void dump ()
  {
    ACE.DEBUG ("-------- module links --------");

    for (Module mp = this.streamHead_; ; mp = mp.next ())
      {
	ACE.DEBUG ("module name = " + mp.name ());
	if (mp == this.streamTail_)
	  break;
      }
    
    ACE.DEBUG ("-------- writer links --------");
    
    Task tp;
    
    for (tp = this.streamHead_.writer (); ; tp = tp.next ())
      {
	ACE.DEBUG ("writer queue name = " + tp.name ());
	tp.dump ();
	ACE.DEBUG ("-------\n");
	if (tp == this.streamTail_.writer ()
	    || (this.linkedUs_ != null && tp == this.linkedUs_.streamHead_.reader ()))
	  break;
      }

    ACE.DEBUG ("-------- reader links --------\n");
    for (tp = this.streamTail_.reader (); ; tp = tp.next ())
      {
	ACE.DEBUG ("reader queue name = " + tp.name ());
	tp.dump ();
	ACE.DEBUG ("-------\n");
	if (tp == this.streamHead_.reader ()
	    || (this.linkedUs_ != null && tp == this.linkedUs_.streamHead_.writer ()))
	  break;
      }
  }

  Module streamHead_ = null;
  // Pointer to the head of the stream.

  Module streamTail_ = null;
  // Pointer to the tail of the stream.

  Stream linkedUs_ = null;
  // Pointer to an adjoining linked stream.

  // = Synchronization objects used for thread-safe streams.
  //  ACE_SYNCH_MUTEX lock_;
  // Protect the stream against race conditions.
  
  //  ACE_SYNCH_CONDITION final_close_;
  // Use to tell all threads waiting on the close that we are done.

}


